File: LENTABP10/src/main/java/com/lenta/lentabp10/models/memory/MemoryTaskWriteOffReasonRepository.java
package com.lenta.lentabp10.models.memory;

import com.lenta.lentabp10.models.repositories.ITaskWriteOffReasonRepository;
import com.lenta.lentabp10.models.task.TaskWriteOffReason;
import com.lenta.shared.models.core.IProduct;

import java.util.ArrayList;
import java.util.List;

public class MemoryTaskWriteOffReasonRepository implements ITaskWriteOffReasonRepository {

    private final List<TaskWriteOffReason> arrWriteOffReason = new ArrayList<>();

    public MemoryTaskWriteOffReasonRepository() {
    }

    @Override
    public List<TaskWriteOffReason> findWriteOffReasonsOfProduct(IProduct product) {
        if (product == null)
        {
            throw new NullPointerException("product");
        }

        List<TaskWriteOffReason> foundWriteOffReason = new ArrayList<>();
        for(int i=0; i<arrWriteOffReason.size(); i++) {
            if ( product.getMaterialNumber() == arrWriteOffReason.get(i).getMaterialNumber() ) {
                foundWriteOffReason.add(arrWriteOffReason.get(i));
            }
        }
        return foundWriteOffReason;
    }

    @Override
    public boolean addWriteOffReason(TaskWriteOffReason writeOffReason) {
        if (writeOffReason == null)
        {
            throw new NullPointerException("writeOffReason");
        }

        int index = -1;
        for(int i=0; i<arrWriteOffReason.size(); i++) {
            if ( writeOffReason.getMaterialNumber() == arrWriteOffReason.get(i).getMaterialNumber()
                    && writeOffReason.getWriteOffReason().getCode() == arrWriteOffReason.get(i).getWriteOffReason().getCode() ) {
                index = i;
            }
        }

        if (index == -1)
        {
            arrWriteOffReason.add(writeOffReason);
            return true;
        }
        return false;
    }

    @Override
    public boolean deleteWriteOffReason(TaskWriteOffReason writeOffReason) {
        if (writeOffReason == null)
        {
            throw new NullPointerException("writeOffReason");
        }

        int index = -1;
        for(int i=0; i<arrWriteOffReason.size(); i++) {
            if ( writeOffReason.getMaterialNumber() == arrWriteOffReason.get(i).getMaterialNumber()
                    && writeOffReason.getWriteOffReason().getCode() == arrWriteOffReason.get(i).getWriteOffReason().getCode() ) {
                index = i;
            }
        }

        if (index == -1)
        {
            return false;
        }
        arrWriteOffReason.remove(index);
        return true;
    }

    @Override
    public boolean deleteWriteOffReasonsForProduct(IProduct product) {
        if (product == null)
        {
            throw new NullPointerException("product");
        }

        List<TaskWriteOffReason> delWriteOffReason = new ArrayList<>();
        for(int i=0; i<arrWriteOffReason.size(); i++) {
            if ( product.getMaterialNumber() == arrWriteOffReason.get(i).getMaterialNumber() ) {
                delWriteOffReason.add(arrWriteOffReason.get(i));
            }
        }

        if ( delWriteOffReason.isEmpty() )
        {
            return false;
        }

        arrWriteOffReason.removeAll(delWriteOffReason);
        return true;
    }

    @Override
    public void clear() {
        arrWriteOffReason.clear();
    }

    @Override
    public TaskWriteOffReason get(int index) {
        return arrWriteOffReason.get(index);
    }

    @Override
    public int lenght() {
        return arrWriteOffReason.size();
    }
}

File: writeoff_breakage/src/main/java/com/lenta/bp10/models/memory/MemoryTaskWriteOffReasonRepository.kt
package com.lenta.bp10.models.memory

import com.lenta.bp10.models.repositories.ITaskWriteOffReasonRepository
import com.lenta.bp10.models.task.TaskWriteOffReason
import com.lenta.shared.models.core.ProductInfo
import java.util.*

class MemoryTaskWriteOffReasonRepository(
        private val writeOffReasons: ArrayList<TaskWriteOffReason> = ArrayList()
) : ITaskWriteOffReasonRepository {

    override fun getWriteOffReasons(): List<TaskWriteOffReason> {
        return writeOffReasons
    }

    override fun findWriteOffReasonsOfProduct(product: ProductInfo): List<TaskWriteOffReason> {
        val foundWriteOffReason = ArrayList<TaskWriteOffReason>()
        for (i in writeOffReasons.indices) {
            if (product.materialNumber == writeOffReasons[i].materialNumber) {
                foundWriteOffReason.add(writeOffReasons[i])
            }
        }
        return foundWriteOffReason
    }

    override fun addWriteOffReason(writeOffReason: TaskWriteOffReason): Boolean {
        var index = -1
        for (i in this.writeOffReasons.indices) {
            if (writeOffReason.materialNumber == this.writeOffReasons[i].materialNumber && writeOffReason.writeOffReason.code == this.writeOffReasons[i].writeOffReason.code) {
                index = i
            }
        }

        if (index == -1) {
            this.writeOffReasons.add(writeOffReason)
            return true
        }
        return false
    }

    override fun deleteWriteOffReason(writeOffReason: TaskWriteOffReason): Boolean {
        var index = -1
        for (i in this.writeOffReasons.indices) {
            if (writeOffReason.materialNumber == this.writeOffReasons[i].materialNumber && writeOffReason.writeOffReason.code == this.writeOffReasons[i].writeOffReason.code) {
                index = i
            }
        }

        if (index == -1) {
            return false
        }
        this.writeOffReasons.removeAt(index)
        return true
    }

    override fun deleteWriteOffReasonsForProduct(product: ProductInfo): Boolean {
        val delWriteOffReason = ArrayList<TaskWriteOffReason>()
        for (i in writeOffReasons.indices) {
            if (product.materialNumber == writeOffReasons[i].materialNumber) {
                delWriteOffReason.add(writeOffReasons[i])
            }
        }

        if (delWriteOffReason.isEmpty()) {
            return false
        }

        writeOffReasons.removeAll(delWriteOffReason)
        return true
    }

    override fun clear() {
        writeOffReasons.clear()
    }

    override fun get(index: Int): TaskWriteOffReason {
        return writeOffReasons[index]
    }

    override fun lenght(): Int {
        return writeOffReasons.size
    }

}